### ISSUE TITLE
Dicts with tuple keys not working in 0.52.0

### ISSUE THREAD

**Issue by kartiksubbarao** [numba/numba#6570]
I’m using a dict with a tuple as the key which worked fine in 0.51.2, but does not work in 0.52.0:

```
import numpy as np
import numba

@numba.njit
def testdict(tuptype):
    tdict = numba.typed.Dict.empty(tuptype, numba.core.types.float32)
    t1 = np.array([3], dtype=np.uint64)
    t2 = np.array([5.67], dtype=np.float32)
    v1 = np.array([10.23], dtype=np.float32)
    tdict[(t1[0], t2[0])] = v1[0]
    # 0.51.2 prints True as expected; 0.52.0 prints False
    print((t1[0], t2[0]) in tdict)

testdict(
  numba.core.types.Tuple((numba.core.types.uint64, numba.core.types.float32)))
```
When I try to check membership of an element that I just stored, 0.52.0 returns False.

**Comment by stuartarchibald** [comment]
Thanks for the report. When I run this I see:
```
NumbaTypeSafetyWarning: unsafe cast from Tuple(int64, float64) to Tuple(uint64, float32). Precision may be lost.
```
which hints at the problem, and I'm afraid it's somewhat subtle. The issue is on this line:
```
tdict[(t1[0], t2[0])] = v1[0]
```
when Numba indexes into the arrays, the returned type is a python scalar, not a NumPy scalar. There's numerous issues about this (e.g. https://github.com/numba/numba/issues/5990) and it's waiting on a somewhat involved fix. The above results in a dictionary key being a tuple of an `int64` and a `float64` which then leads to type casting and ends up with something invalid.

A way to get what I think you are looking for is this:
```python
import numpy as np
import numba

kty = numba.core.types.Tuple((numba.core.types.uint64, numba.core.types.float32))

@numba.njit
def testdict():
    tdict = numba.typed.Dict.empty(kty, numba.core.types.float32)
    t1 = np.array([3], dtype=np.uint64)
    t2 = np.array([5.67], dtype=np.float32)
    v1 = np.array([10.23], dtype=np.float32)
    tdict[(np.uint64(t1[0]), np.float32(t2[0]))] = v1[0]
    print((np.uint64(t1[0]), np.float32(t2[0])) in tdict)

testdict()
```

**Comment by kartiksubbarao** [comment]
Thanks for the explanation @stuartarchibald. Unfortunately the workaround doesn't seem to work for me -- it still prints False (and I didn't get the NumbaTypeSafetyWarning message with the original code). I'm wondering if I'm running a different version than you are. I installed numpy and numba via pip install. Here are the versions on my system (Fedora 32):

# np.version.version
Numpy: numpy 1.19.4

# numba._version.get_versions()
{'version': '0.52.0', 'full': '18825058a6e32bb92f3941f3c8d14a818f19279f'}

**Comment by sklam** [comment]
Looks like this has to do with the new refprune pass

**Comment by sklam** [comment]
Yup, running the reproducer under `NUMBA_LLVM_REFPRUNE_PASS=0` stops it from failing.

**Comment by sklam** [comment]
Update: I'm ruling out this bug being caused by the new refprune pass. I am finding that on 0.51.0 (3a14664a), the reproducer shows:

```bash
$ NUMBA_OPT=0 python iss6570.py
False
$ NUMBA_OPT=1 python iss6570.py
False
$ NUMBA_OPT=3 python iss6570.py
True
```

I have also disabled the old refpruner. Resulting changing in optimization seems to imply that we relying on an undefined behavior.

**Comment by sklam** [comment]
I have found the true cause of this issue---a problem with garbage value in alignment padding bytes. The key type is a `{i64, float32}`, which has an 8-byte alignment. The true data is an 8-byte (i64) + a 4-byte (float32) resulting in 12-byte data structure. So the compiler puts 4-byte padding to make it 8-byte aligned. The dict container C code is told the key type is 16-byte long (from sizeof) and does `memcmp()` to find matching keys. The failure was caused by that 4-byte containing random garbage.
